### PR TITLE
Annotations

### PR DESCRIPTION
**Pull request by elijahbenizzy**
Adds annotations workflow. See #285.

High-level:
1. You can annotate an action
2. This can be associated with a number of "observations", each one on a piece of data
3. You can add notes/ground truths/thumbs up/thumbs down on these
4. There is a project-level view (meant to be a queue) that you can work on
5. You can download these into a CSV.

## Changes
- BE -- added endpoints to create, expose
- FE -- added annotations view to app
- FE -- added annotations by project
- Edit + create modals to both

## How I tested this
- manually

## Notes

Remaining:
- [ ] Ensure this doesn't break the S3 implementation
- [ ] Add docs + appropriate references

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

----

> [!IMPORTANT]
> Add comprehensive annotations workflow with backend endpoints and frontend views for creating, updating, and managing annotations.
> 
>   - **Backend**:
>     - Add `AnnotationsBackendMixin` to `backend.py` for handling annotations.
>     - New endpoints in `run.py` for creating, updating, and retrieving annotations.
>   - **Frontend**:
>     - Add `AnnotationsViewContainer` and `AnnotationsView` in `AppView.tsx` for displaying annotations.
>     - Add `Drawer` component in `drawer.tsx` for annotation editing.
>     - Update `ProjectList.tsx` to include annotations link.
>   - **API**:
>     - Add models `AnnotationCreate`, `AnnotationUpdate`, `AnnotationOut`, `AnnotationObservation`, and `AnnotationDataPointer`.
>     - Update `DefaultService.ts` with new annotation-related API calls.
>   - **Misc**:
>     - Update `package.json` dependencies for new features.
>     - Add `Chip` component updates in `chip.tsx` for annotation tags.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=DAGWorks-Inc%2Fburr&utm_source=github&utm_medium=referral)<sup> for bcd21d7c2dea46c6a82264e1b15998576f5dad4f. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
TODO, will add more:
- [ ] skip loading of annotations if BE does not support it
- [ ] Fix up so it works with spans
    - [ ] Add in UI
    - [ ] Ensure it's in data model
- [ ] Wire through to BE
- [ ] Test on S3/implement

**Comment by elijahbenizzy** [comment]
Annotations view with inline app
image

Annotations edit/create window: 
image

Inline annotation edit modal 
image

File: burr/tracking/server/backend.py [hunk 1]
```diff
@@ -197,6 +260,93 @@ class LocalBackend(BackendBase):
...
+    async def _load_project_annotations(self, project_id: str):
+        annotations_path = self._get_annotation_path(project_id)
+        annotations = []
+        if os.path.exists(annotations_path):
+            async with aiofiles.open(annotations_path) as f:
+                for line in await f.readlines():
+                    annotations.append(AnnotationOut.parse_raw(line))
+        return annotations
+
+    async def create_annotation(
```
Comment (elijahbenizzy) [inline burr/tracking/server/backend.py:275]: Add comments

File: burr/tracking/server/backend.py [hunk 2]
```diff
@@ -197,6 +260,93 @@ class LocalBackend(BackendBase):
...
+            created=datetime.now(),
+            updated=datetime.now(),
+            **annotation.dict(),
+        )
+        annotations_path = self._get_annotation_path(project_id)
+        async with aiofiles.open(annotations_path, "a") as f:
+            await f.write(annotation_out.json() + "\n")
+        return annotation_out
+
+    async def update_annotation(
```
Comment (elijahbenizzy) [inline burr/tracking/server/backend.py:302]: comments

File: burr/tracking/server/backend.py [hunk 3]
```diff
@@ -197,6 +260,93 @@ class LocalBackend(BackendBase):
...
+                status_code=404,
+                detail=f"Annotation: {annotation_id} from project: {project_id} not found",
+            )
+        annotations_path = self._get_annotation_path(project_id)
+        async with aiofiles.open(annotations_path, "w") as f:
+            for a in all_annotations:
+                await f.write(a.json() + "\n")
+        return annotation_out
+
+    async def get_annotations(
```
Comment (elijahbenizzy) [inline burr/tracking/server/backend.py:328]: comments

File: burr/tracking/server/run.py [hunk 4]
```diff
@@ -217,6 +227,58 @@ async def get_application_logs(
     )
 
 
+@app.post(
+    "/api/v0/{project_id}/{app_id}/{partition_key}/{sequence_id}annotations",
```
Comment (elijahbenizzy) [inline burr/tracking/server/run.py:231]: Add another `/` here

File: burr/tracking/server/run.py [hunk 5]
```diff
@@ -217,6 +227,58 @@ async def get_application_logs(
...
+    annotation_id: int,
+    annotation: AnnotationUpdate,
+):
+    return await backend.update_annotation(
+        annotation_id=annotation_id, annotation=annotation, project_id=project_id
+    )
+
+
+@app.get("/api/v0/{project_id}/annotations", response_model=Sequence[AnnotationOut])
+async def get_annotations(
```
Comment (elijahbenizzy) [inline burr/tracking/server/run.py:267]: Make it empty if backend doesn't support it

File: burr/tracking/server/run.py [hunk 6]
```diff
@@ -217,6 +227,58 @@ async def get_application_logs(
     )
 
 
+@app.post(
+    "/api/v0/{project_id}/{app_id}/{partition_key}/{sequence_id}annotations",
+    response_model=AnnotationOut,
+)
+async def create_annotation(
```
Comment (elijahbenizzy) [inline burr/tracking/server/run.py:234]: Empty if BE doesn't support it

File: burr/tracking/server/schema.py [hunk 7]
```diff
@@ -182,3 +182,52 @@ class BackendSpec(pydantic.BaseModel):
     indexing: bool
     snapshotting: bool
     supports_demos: bool
+    supports_annotations: bool
+
+
+class AnnotationDataPointer(pydantic.BaseModel):
```
Comment (elijahbenizzy) [inline burr/tracking/server/schema.py:188]: Comments on these

File: telemetry/ui/src/components/common/chip.tsx [hunk 8]
```diff
@@ -29,22 +33,63 @@ export type ChipType = keyof typeof chipColorMap;
  * This makes it easy to centralize meaning across the repository
  * although it breaks encapsulation.
  */
+// export const Chip = (props: {
```
Comment (elijahbenizzy) [inline telemetry/ui/src/components/common/chip.tsx:36]: delete

File: telemetry/ui/src/components/routes/app/StepList.tsx [hunk 9]
```diff
@@ -1222,9 +1275,9 @@ export const StepList = (props: {
                 <MinimizeTableIcon
                   className={`h-4 w-4 ${props.fullScreen ? 'text-gray-300' : 'hover:scale-110 cursor-pointer text-gray-600'} ${props.allowMinimized ? '' : 'invisible'}`}
                   onClick={(e) => {
-                    if (props.fullScreen) {
-                      return;
-                    }
+                    // if (props.fullScreen) {
```
Comment (elijahbenizzy) [inline telemetry/ui/src/components/routes/app/StepList.tsx:1278]: delete

**Review by skrawcz — APPROVED** [review]
this looks good for a v1

**Pull request merged**